flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java

```diff
@@ -0,0 +1,429 @@
+
+	/** Time interval (ms) to run the cleanup task; also used as the default TTL. */
+	private final long cleanupInterval;
+
+	private final Timer cleanupTimer;
+
+	/**
+	 * Instantiates a new cache for permanent BLOBs which are also available in an HA store.
+	 *
+	 * @param serverAddress
+	 * 		address of the {@link BlobServer} to use for fetching files from
+	 * @param blobClientConfig
+	 * 		global configuration
+	 * @param blobView
+	 * 		(distributed) HA blob store file system to retrieve files from first
+	 *
+	 * @throws IOException
+	 * 		thrown if the (local or distributed) file storage cannot be created or is not usable
+	 */
+	public PermanentBlobCache(
+		final InetSocketAddress serverAddress,
+		final Configuration blobClientConfig,
+		final BlobView blobView) throws IOException {
+
+		this.serverAddress = checkNotNull(serverAddress);
+		this.blobClientConfig = checkNotNull(blobClientConfig);
+		this.blobView = checkNotNull(blobView, "blobStore");
+		this.readWriteLock = new ReentrantReadWriteLock();
+
+		// configure and create the storage directory
+		String storageDirectory = blobClientConfig.getString(BlobServerOptions.STORAGE_DIRECTORY);
+		this.storageDir = BlobUtils.initLocalStorageDirectory(storageDirectory);
+		LOG.info("Created permanent BLOB cache storage directory " + storageDir);
+
+		// configure the number of fetch retries
+		final int fetchRetries = blobClientConfig.getInteger(BlobServerOptions.FETCH_RETRIES);
+		if (fetchRetries >= 0) {
+			this.numFetchRetries = fetchRetries;
+		} else {
+			LOG.warn("Invalid value for {}. System will attempt no retries on failed fetch operations of BLOBs.",
+				BlobServerOptions.FETCH_RETRIES.key());
+			this.numFetchRetries = 0;
+		}
+
+		// Initializing the clean up task
+		this.cleanupTimer = new Timer(true);
+
+		this.cleanupInterval = blobClientConfig.getLong(BlobServerOptions.CLEANUP_INTERVAL) * 1000;
+		this.cleanupTimer.schedule(this, cleanupInterval, cleanupInterval);
+
+		// Add shutdown hook to delete storage directory
+		shutdownHook = BlobUtils.addShutdownHook(this, LOG);
+	}
+
+	/**
+	 * Registers use of job-related BLOBs.
+	 * <p>
+	 * Using any other method to access BLOBs, e.g. {@link #getHAFile}, is only valid within calls
+	 * to <tt>registerJob(JobID)</tt> and {@link #releaseJob(JobID)}.
+	 *
+	 * @param jobId
+	 * 		ID of the job this blob belongs to
+	 *
+	 * @see #releaseJob(JobID)
+	 */
+	public void registerJob(JobID jobId) {
+		synchronized (jobRefCounters) {
+			RefCount ref = jobRefCounters.get(jobId);
+			if (ref == null) {
+				ref = new RefCount();
+				jobRefCounters.put(jobId, ref);
+			} else {
+				// reset cleanup timeout
+				ref.keepUntil = -1;
+			}
+			++ref.references;
+		}
+	}
+
+	/**
+	 * Unregisters use of job-related BLOBs and allow them to be released.
+	 *
+	 * @param jobId
+	 * 		ID of the job this blob belongs to
+	 *
+	 * @see #registerJob(JobID)
+	 */
+	public void releaseJob(JobID jobId) {
```

Comment:
With the current code, it wouldn't have mattered, whether the counter is below zero since the cleanup logic inside `PermanentBlobCache#run()` accounted for that and `releaseJob` only set the cleanup time once. It should, however, also considered an improper use case and we may guard against it.

Reply:
I was just wondering, because what happens if it is negative and then this job is registered again? Will the counter then be at `<= 0`? This would mean that all blobs associated with this job will be deleted at the next clean up cycle, right?

Reply:
This behaviour changed with the last commits - the reference counter can now not go below 0 and issues a warning in case it would.